axlearn/common/trainer.py

```diff
@@ -376,22 +382,46 @@ def _stop_watchdog(self):
     def _watchdog_loop(self, *, context_stack: list[InvocationContext]):
         cfg = self.config
         install_context_stack(context_stack)
+        time_elapsed_in_sec_since_last_check: float = 0.0
+        # Set a scanning time to 10 mins or the watchdog_timeout_seconds, whichever is smaller.
+        health_check_in_sec = min(cfg.watchdog_timeout_seconds, 600)
         while True:
             last_step = self.step
-            if self._watchdog_stopping.wait(timeout=cfg.watchdog_timeout_seconds):
+            if self._watchdog_stopping.wait(health_check_in_sec):
                 break
             current_step = self.step
             if current_step == last_step:
-                self._step_log(
-                    "Watchdog triggered because step has not incremented in the last %s seconds.\n"
-                    "NOTE: this is not an error message, but meant to help debugging "
-                    "in case the trainer is stuck.\n"
-                    "Threads:\n%s",
-                    cfg.watchdog_timeout_seconds,
-                    "\n".join(itertools.chain.from_iterable(thread_stack_traces())),
-                )
+                time_elapsed_in_sec_since_last_check += health_check_in_sec
+                # When device_monitor is enabled, we can check if the host is idle
+                # and trigger the watchdog proactively.
+                if self._device_monitor is not None:
+                    if self._device_monitor.is_host_idle():
+                        self._step_log(
+                            "Watchdog triggered because step has not incremented in the last %s "
+                            "seconds and the host is idle.\n"
+                            "NOTE: this is not an error message, but meant to help debugging "
+                            "in case the trainer is stuck.\n"
+                            "Threads:\n%s",
+                            time_elapsed_in_sec_since_last_check,
+                            "\n".join(
+                                itertools.chain.from_iterable(thread_stack_traces()),
+                            ),
+                        )
+                        # TODO(kelvin-zou): maybe crash the program here?
```

Comment:
+1 on crashing the program to avoid the waste. But also making it configurable so that a job can configure how often the health check should run, and when to crash (like if idle for x hours)

Reply:
Crashing the program here would trigger a JobSet restart (all Pods would be recreated). Are there some failure modes that this would address?

Reply:
@nstogner  atm it is only logging faster, so that our log parser can catch it earlier and report to GCP. let's chat offline about how to crash here, the logic is much more complex than an `raise Error`.

Reply:
> Crashing the program here would trigger a JobSet restart (all Pods would be recreated). Are there some failure modes that this would address?


When a job gets stuck, restarting the jobset can sometimes help (but not always)